tests/test_image.py

```diff
@@ -34,73 +38,271 @@ def test_input_array_not_mutated(image_type: type[ImageType]):
+    # Make sure there's a value to mask in the input array
+    a = np.array([[[np.nan]]])
+    img = wrap_image(a, type=image_type)
+    image = Image.from_image(img, nodata_input=0)
+
+    output_nodata, output_dtype = val_dtype
+    with pytest.raises(ValueError, match="does not fit in the array dtype"):
+        # Unwrap to force computation for lazy arrays
+        unwrap_image(
+            image.apply_ufunc_across_bands(
+                lambda x: np.ones_like(x).astype(output_dtype),
+                nodata_output=output_nodata,
+                skip_nodata=skip_nodata,
+                output_dims=[["variable"]],
+                output_sizes={"variable": a.shape[0]},
+                output_dtypes=[a.dtype],
+            )
+        )
+
+
+@parametrize_image_types()
+@pytest.mark.parametrize("skip_nodata", [True, False])
+@pytest.mark.parametrize(
+    "val_dtypes", [(-1, np.uint8, np.int8), (np.nan, np.int16, np.float64)]
+)
+def test_nodata_output_with_allow_cast(
+    val_dtypes: tuple[int | float, np.dtype, np.dtype],
+    image_type: type[ImageType],
+    skip_nodata: bool,
+):
+    """Test that an unsupported nodata_output value correctly casts if allowed."""
+    # Make sure there's a value to mask in the input array
+    a = np.array([[[np.nan]]])
+    img = wrap_image(a, type=image_type)
+    image = Image.from_image(img, nodata_input=0)
+
+    output_nodata, output_dtype, expected_dtype = val_dtypes
+    # Unwrap to force computation for lazy arrays
+    result = unwrap_image(
+        image.apply_ufunc_across_bands(
+            lambda x: np.ones_like(x).astype(output_dtype),
+            nodata_output=output_nodata,
+            skip_nodata=skip_nodata,
+            allow_cast=True,
+            output_dims=[["variable"]],
+            output_sizes={"variable": a.shape[0]},
+            output_dtypes=[a.dtype],
+        )
+    )
 
-    image = Image.from_image(wrap_image(a, type=image_type))
+    assert result.dtype == expected_dtype
+
+
+@pytest.mark.parametrize("nodata_output", [np.nan, 42.0])
+@parametrize_image_types()
+@pytest.mark.parametrize("skip_nodata", [True, False])
+def test_nodata_output_set(
+    nodata_output: int | float, image_type: type[ImageType], skip_nodata: bool
+):
+    """Test that NoData in the image are filled or not."""
+    nodata_input = 0
 
-    output = image.apply_ufunc_across_bands(
-        func=lambda x: x,
-        nan_fill=nan_fill,
-        # Masking NoData would broadcast NaNs across the band dimension since a missing
-        # value in any band is considered missing in all bands of the output, so skip
-        # that step.
-        mask_nodata=False,
+    # Encoded NoData and NaN should both be replaced across bands with the nodata_output
+    # value.
+    a = np.array([[[nodata_input, 1, np.nan]]])
+    expected_output = np.array([[[nodata_output, 1, nodata_output]]])
+
+    image = Image.from_image(wrap_image(a, type=image_type), nodata_input=0)
```

Comment:
Nit.  If you declare `nodata_input` and it doesn't change, should you use it on line 117 (as well as line 114)?